File: app/agent/trace.py
"""
Agent Trace：记录 Agent 执行过程的每一步

Trace 就是 Agent 的黑匣子：每一次 LLM 决策、每一次工具调用、
每一个参数、每一条结果，全部留痕。
"""

from dataclasses import dataclass, field, asdict


@dataclass
class TraceStep:
    """
    一次工具调用步骤。
    """
    round: int                  # 第几轮循环（1 开始）
    tool_name: str              # 调用了哪个工具
    tool_arguments: dict        # 传入的参数
    tool_result: dict           # 工具的返回结果（含错误时也是 dict）
    response_time: float = 0.0  # 本轮 LLM 调用耗时

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class AgentTrace:
    """一次完整 Agent 任务的执行记录。"""
    user_input: str             # 用户输入
    steps: list = field(default_factory=list)   # 工具调用步骤列表
    final_answer: str = ""      # 最终回答
    success: bool = False       # 任务是否成功
    total_time: float = 0.0     # 任务总耗时（秒）
    total_tokens: int = 0       # 任务总 token 消耗
    num_llm_calls: int = 0      # LLM 被调用了多少次（= 循环轮数）
    error: str = ""             # 失败原因（成功时为空）

    def to_dict(self) -> dict:
        """转成 JSON 可序列化字典 —— 后面写测试报告 / Web 展示都靠它。"""
        return {
            "user_input": self.user_input,
            "steps": [s.to_dict() for s in self.steps],
            "final_answer": self.final_answer,
            "success": self.success,
            "total_time": self.total_time,
            "total_tokens": self.total_tokens,
            "num_llm_calls": self.num_llm_calls,
            "num_tool_calls": len(self.steps),
            "error": self.error,
        }

    def pretty_print(self):
        """终端友好展示。"""
        print(f"  用户输入     : {self.user_input}")
        for s in self.steps:
            print(f"  第 {s.round} 轮 -> {s.tool_name}({s.tool_arguments})")
            print(f"     工具结果 : {s.tool_result}")
        print(f"  最终回答     : {self.final_answer}")
        print(f"  任务状态     : {'成功' if self.success else '失败'}")
        print(f"  总耗时       : {self.total_time}s | LLM 调用 {self.num_llm_calls} 次 "
              f"| 工具调用 {len(self.steps)} 次 | 总 token {self.total_tokens}")

File: app/web/static/trace.js
/**
 * Trace 可视化渲染器
 *
 * 把 Agent Trace 的 JSON 数据渲染成步骤化的 HTML：
 *   用户输入 → [第1轮] 工具调用 → [第2轮] 工具调用 → 最终回答
 *
 * 在 dashboard.html（弹窗）和 case_detail.html（内嵌）中共享使用。
 */

/**
 * 渲染完整的 Trace HTML
 * @param {Object} trace - trace.to_dict() 返回的对象
 * @returns {string} HTML 字符串
 */
function renderTrace(trace) {
    if (!trace || !trace.steps) {
        return '<p class="trace-empty">暂无 Trace 数据</p>';
    }

    let html = '<div class="trace-flow">';

    // ---- 用户输入 ----
    html += `
        <div class="trace-step trace-step-user">
            <div class="trace-step-icon">👤</div>
            <div class="trace-step-content">
                <div class="trace-step-title">用户输入</div>
                <div class="trace-step-body">${escapeHtml(trace.user_input)}</div>
            </div>
        </div>
    `;

    // ---- 每轮工具调用 ----
    trace.steps.forEach(function(step) {
        var isBlocked = step.tool_result && step.tool_result.blocked;
        var isError = step.tool_result && step.tool_result.error && !isBlocked;
        var stepClass = isBlocked ? 'trace-step-blocked' : (isError ? 'trace-step-error' : 'trace-step-tool');
        var icon = isBlocked ? '🚫' : (isError ? '⚠️' : '🔧');
        var statusLabel = isBlocked ? ' (权限拦截)' : (isError ? ' (执行错误)' : '');

        html += `
            <div class="trace-arrow">↓</div>
            <div class="trace-step ${stepClass}">
                <div class="trace-step-icon">${icon}</div>
                <div class="trace-step-content">
                    <div class="trace-step-title">
                        第 ${step.round} 轮 · ${escapeHtml(step.tool_name)}${statusLabel}
                    </div>
                    <div class="trace-step-body">
                        <div class="trace-section">
                            <span class="trace-label">参数</span>
                            <pre class="trace-code">${escapeHtml(JSON.stringify(step.tool_arguments, null, 2))}</pre>
                        </div>
                        <div class="trace-section">
                            <span class="trace-label">结果</span>
                            <pre class="trace-code">${escapeHtml(JSON.stringify(step.tool_result, null, 2))}</pre>
                        </div>
                    </div>
                </div>
            </div>
        `;
    });

    // ---- 最终回答 ----
    if (trace.final_answer) {
        var icon = trace.success ? '✅' : '❌';
        var stepClass = trace.success ? 'trace-step-final' : 'trace-step-error';
        var label = trace.success ? '最终回答 (成功)' : '最终回答 (失败)';
        html += `
            <div class="trace-arrow">↓</div>
            <div class="trace-step ${stepClass}">
                <div class="trace-step-icon">${icon}</div>
                <div class="trace-step-content">
                    <div class="trace-step-title">${label}</div>
                    <div class="trace-step-body trace-answer-box">${escapeHtml(trace.final_answer)}</div>
                </div>
            </div>
        `;
    }

    html += '</div>';

    // ---- 统计信息 ----
    var numToolCalls = trace.num_tool_calls || trace.steps.length;
    html += `
        <div class="trace-stats">
            <span class="trace-stat-item">LLM 调用: <strong>${trace.num_llm_calls}</strong> 次</span>
            <span class="trace-stat-item">工具调用: <strong>${numToolCalls}</strong> 次</span>
            <span class="trace-stat-item">总耗时: <strong>${trace.total_time}</strong>s</span>
            <span class="trace-stat-item">总 Token: <strong>${trace.total_tokens}</strong></span>
        </div>
    `;

    return html;
}

/**
 * HTML 转义，防止用户输入的内容破坏 HTML 结构
 */
function escapeHtml(text) {
    if (text == null) return '';
    var div = document.createElement('div');
    div.textContent = String(text);
    return div.innerHTML;
}
